### DISCUSSION TITLE
Compiling rust-horizon on Windows

### DISCUSSION THREAD

**Discussion by nemui** [AzureMarker/rust-horizon#6] [Q&A] [answered]
Hello!
I was very happy to discover rust-horizon and related projects after tinkering with https://github.com/rust3ds for a couple of evenings.

Tried compiling on Windows following the guide at https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html, but running `x.py build -i library/std` results in an error:
```
Building stage0 std artifacts (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc)
   Compiling compiler_builtins v0.1.72
error[E0412]: cannot find type `c_char` in module `core::ffi`
  --> C:\Users\metap\.cargo\registry\src\github.com-1ecc6299db9ec823\compiler_builtins-0.1.72\src\mem\mod.rs:74:58
   |
74 |     pub unsafe extern "C" fn strlen(s: *const core::ffi::c_char) -> usize {
   |                                                          ^^^^^^ not found in `core::ffi`

For more information about this error, try `rustc --explain E0412`.
error: could not compile `compiler_builtins` due to previous error
```

Building `master` seems to be working fine, perhaps `horizon-std` needs some additional arguments/special config values? 
Any pointers would be greatly appreciated!

**Comment by Meziu** [comment, marked as answer]
Right now no branch is fully usable, as we are preparing to push our changes to Rust’s master branch (which would make building much easier). 

Still, if you want to try `horizon-std` with the examples we have in [ctru-rs](https://github.com/Meziu/ctru-rs) you will need to install the full `3ds-dev` module from `devkitPRO`’s toolchain and build the stage2 rustc compiler (from `rust-horizon`). At that point you should add the `rustc` keyword (under the `[build]` category) in the `.cargo/config.toml` file and set it to the custom `rustc` you just built. Then you can use our [cargo-3ds](https://github.com/Meziu/cargo-3ds) tool like a normal cargo tool and you are good to go.

Since neither our `std`, nor `cargo-3ds`, nor `ctru-rs` are released, the process is very cumbersome and requires the cloning of multiple repositories, plus some tinkering with the toolchain. Once everything is published, the process will look something like:
- Install `devkitPRO`‘s toolchain
- `cargo install cargo-3ds`
- `cargo 3ds run`

We are getting there though, so check out our progress 😃

**Reply by nemui** [reply]
Thank you for the detailed answer! I'd like to try the examples from `ctru-rs`, got `devkitPRO`/`3ds-dev` installed and sample programs running in Citra. Building rustc from `rust-horizon` is giving me trouble though; perhaps it's better to wait for the published versions of your `std`, `cargo-3ds` and `ctru-rs` and learn `libctru` in the meantime.
Looking forward to the published crates, and again - amazing work!

**Reply by Meziu** [reply]
You can have a look at `libctru`‘s examples in the meantime, knowing what’s underneath is very useful. Also, we aim to port most of those over, so you can compare the Rust code to the C code and see what’s going on.

**Reply by AzureMarker** [reply]
@nemui The most stable branches to use in this repo are `feature/horizon-std` and `feature/horizon-threads` (the first, plus thread support).

**Reply by AzureMarker** [reply]
Additionally, you can add your local rustc install to rustup to utilize the (more common) `rust-toolchain` file, or use `rustup toolchain set override`.

See https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html#creating-a-rustup-toolchain
